internal/vector_clock.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2008-2018, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	. "github.com/hazelcast/hazelcast-go-client/internal/protocol"
+	"github.com/hazelcast/hazelcast-go-client/serialization"
+	"sync"
+)
+
+const (
+	FId                = 0
+	VectorClockClassId = 43
+)
+
+type VectorClock struct {
+	mutex             sync.RWMutex //guards replicaTimestamps
+	replicaTimestamps map[*string]int64
+}
+
+func NewVectorClock() *VectorClock {
+	return &VectorClock{replicaTimestamps: make(map[*string]int64)}
+}
+
+func (*VectorClock) FactoryId() int32 {
+	return FId
+}
+
+func (*VectorClock) ClassId() int32 {
+	return VectorClockClassId
+}
+
+func (v *VectorClock) WriteData(output serialization.DataOutput) (err error) {
+	v.mutex.RLock()
+	defer v.mutex.RUnlock()
+	output.WriteInt32(int32(len(v.replicaTimestamps)))
+	for key, value := range v.replicaTimestamps {
+		output.WriteUTF(*key)
+		output.WriteInt64(value)
+	}
+	return
+}
+
+func (v *VectorClock) ReadData(input serialization.DataInput) error {
+	stateSize, err := input.ReadInt32()
+	if err != nil {
+		return err
+	}
+	for i := int32(0); i < stateSize; i++ {
+		replicaId, err := input.ReadUTF()
+		if err != nil {
+			return err
+		}
+		timestamp, err := input.ReadInt64()
+		if err != nil {
+			return err
+		}
+		v.mutex.Lock()
+		v.replicaTimestamps[&replicaId] = timestamp
+		v.mutex.Unlock()
+	}
+	return nil
+}
+
+func (v *VectorClock) EntrySet() (entrySet []*Pair) {
+	v.mutex.RLock()
+	defer v.mutex.RUnlock()
+	entrySet = make([]*Pair, len(v.replicaTimestamps))
+	i := 0
+	for key, value := range v.replicaTimestamps {
+		entrySet[i] = NewPair(key, value)
+		i += 1
+	}
+	return entrySet
+}
+
+func (v *VectorClock) IsAfter(other *VectorClock) (isAfter bool) {
+	anyTimestampGreater := false
+	for replicaId, otherReplicaTimestamp := range other.replicaTimestamps {
+		localReplicaTimestamp := v.TimestampForReplica(replicaId)
+		if localReplicaTimestamp == 0 || localReplicaTimestamp < otherReplicaTimestamp {
+			return false
+		} else if localReplicaTimestamp > otherReplicaTimestamp {
+			anyTimestampGreater = true
+		}
+	}
+	containsAll := true
+	v.mutex.RLock()
+	for replicaId, _ := range v.replicaTimestamps {
+		if _, containsAll = other.replicaTimestamps[replicaId]; !containsAll {
+			break
+		}
+	}
+	v.mutex.RUnlock()
+	return anyTimestampGreater || containsAll
```

Comment:
this check is not true. if two vector clocks are exactly the same, this will return true although it needs to return false. 

You actually do not need the second loop. At this point, it is enough to check if `v` has more entries than `other` because you already know that `v` has all the entries that `other` has by the first loop.